### ISSUE TITLE
[BUG] Fail to Upgrade `harvester-cloud-provider`  on RKE2 cluster (`rancher-v2.8.4`)

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#5873]
**Describe the bug**
Fail to Upgrade `harvester-cloud-provider` on RKE2 (`v1.28`) cluster (`rancher-v2.8.4`)

> [!NOTE]
> * Similiar to harvester/harvester/issues/5382

`0/1 nodes are available: 1 node(s) didn't match pod anti-affinity rules. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod..`
![image](https://github.com/harvester/harvester/assets/2773781/7f408b08-6dea-49ba-8726-cc3391d16249)

**To Reproduce**
1. Setup `harvester-v1.2.2` and `rancher-v2.8.4`
3. Import Harvester to Rancher
4. <details>
   <summary> Create RKE2 cluster (<code>v1.28.9+rke2r1</code>)
   </summary>

   ![image](https://github.com/harvester/harvester/assets/2773781/75ea2fef-c492-4ad3-b979-1d8ba78e43b5)
   `harvester-cloud-provider` and `harvester-csi-driver`
   ![image](https://github.com/harvester/harvester/assets/2773781/6342e214-c61b-486b-a784-976862242fce)
   </details>
5. Upgrade `harvester-cloud-provider` (`0.2.300` -> `103.0.2+up0.2.4`) :x: 
   * Upgrade
     ![image](https://github.com/harvester/harvester/assets/2773781/e956f281-237a-4e74-9759-60d37fa2422d)
     ![image](https://github.com/harvester/harvester/assets/2773781/8a511509-6f93-4ded-9644-5e7eab5cf05a)
   * After
     ![image](https://github.com/harvester/harvester/assets/2773781/158d2d20-8921-416b-b6f0-f7ab4f766b3c)
     `0/1 nodes are available: 1 node(s) didn't match pod anti-affinity rules. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod..`
     ![image](https://github.com/harvester/harvester/assets/2773781/7f408b08-6dea-49ba-8726-cc3391d16249)

     

**Expected behavior**
Should be able to upgrade Rancher harvester app

**Support bundle**
[supportbundle_284FailUpgradeCloudProvider-24T10-50-58Z.zip](https://github.com/harvester/harvester/files/15433246/supportbundle_284FailUpgradeCloudProvider-24T10-50-58Z.zip)

**Rancher Log**
[myrke2-128-pool1-ef6ffd22-7jtc8-2024-05-24_10_51_16.tar.gz](https://github.com/harvester/harvester/files/15433245/myrke2-128-pool1-ef6ffd22-7jtc8-2024-05-24_10_51_16.tar.gz)

**Environment**
* Harvester
  * Version: `v1.2.2`
  * Profile: QEMU/KVM, single nodes (16C/32G/500G)
  * ui-source: `Auto`
* Rancher
  * Version: `v2.8.4`
  * Profile: Helm (`v1.28.9+k3s1`) in QEMU/KVM (2C/4G)
  * RKE2 Version: `v1.28.9 +rke2r1`

**Additional context**
1. [helm-operation-cfns7_undefined.log](https://github.com/harvester/harvester/files/15433169/helm-operation-cfns7_undefined.log)
    ```
    helm upgrade --history-max=5 --install=true --namespace=kube-system --timeout=10m0s --values=/home/shell/helm/values-harvester-cloud-provider-103.0.2-up0.2.4.yaml --version=103.0.2+up0.2.4 --wait=true harvester-cloud-provider /home/shell/helm/harvester-cloud-provider-103.0.2-up0.2.4.tgz
    checking 8 resources for changes
    Looks like there are no changes for ServiceAccount "kube-vip"
    Patch ServiceAccount "harvester-cloud-provider" in namespace kube-system
    Looks like there are no changes for ClusterRole "kube-vip"
    Looks like there are no changes for ClusterRole "harvester-cloud-provider"
    Looks like there are no changes for ClusterRoleBinding "kube-vip"
    Patch ClusterRoleBinding "harvester-cloud-provider" in namespace 
    Patch DaemonSet "kube-vip" in namespace kube-system
    Patch Deployment "harvester-cloud-provider" in namespace kube-system
    beginning wait for 8 resources with timeout of 10m0s
    Deployment is not ready: kube-system/harvester-cloud-provider. 0 out of 1 expected pods are ready
    DaemonSet is not ready: kube-system/kube-vip. 0 out of 1 expected pods have been scheduled
    Deployment is not ready: kube-system/harvester-cloud-provider. 0 out of 1 expected pods are ready
    ...
    2024-05-24T10:45:07.219067445Z Deployment is not ready: kube-system/harvester-cloud-provider. 0 out of 1 expected pods are ready
    2024-05-24T10:45:09.135615116Z Error: UPGRADE FAILED: context deadline exceeded
    ```

2. `rke2-server.service`
    ```
    # journalctl -u rke2-server.service --follow
    May 24 10:13:43 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:43Z" level=error msg="error syncing 'kube-system/rke2-coredns': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-coredns\" not found, requeuing"
    May 24 10:13:44 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:44Z" level=error msg="error syncing 'kube-system/rke2-ingress-nginx': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-ingress-nginx\" not found, requeuing"
    May 24 10:13:45 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:45Z" level=error msg="error syncing 'kube-system/rke2-metrics-server': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-metrics-server\" not found, requeuing"
    May 24 10:13:45 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:45Z" level=error msg="error syncing 'kube-system/rke2-snapshot-controller-crd': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-controller-crd\" not found, requeuing"
    May 24 10:13:46 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:46Z" level=error msg="error syncing 'kube-system/rke2-snapshot-controller': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-controller\" not found, requeuing"
    May 24 10:13:46 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:46Z" level=error msg="error syncing 'kube-system/rke2-snapshot-validation-webhook': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-validation-webhook\" not found, requeuing"
    May 24 10:13:52 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:13:52Z" level=info msg="Adding node myrke2-128-pool1-ef6ffd22-7jtc8-a99bff44 etcd status condition"
    May 24 10:15:45 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:15:45Z" level=info msg="Tunnel authorizer set Kubelet Port 10250"
    May 24 10:22:04 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:22:04Z" level=info msg="Updating TLS secret for kube-system/rke2-serving (count: 11): map[field.cattle.io/projectId:c-m-c79fjcww:p-gz6xh listener.cattle.io/cn-10.43.0.1:10.43.0.1 listener.cattle.io/cn-127.0.0.1:127.0.0.1 listener.cattle.io/cn-192.168.0.52:192.168.0.52 listener.cattle.io/cn-__1-f16284:::1 listener.cattle.io/cn-kubernetes:kubernetes listener.cattle.io/cn-kubernetes.default:kubernetes.default listener.cattle.io/cn-kubernetes.default.svc:kubernetes.default.svc listener.cattle.io/cn-kubernetes.default.svc.cluster.local:kubernetes.default.svc.cluster.local listener.cattle.io/cn-localhost:localhost listener.cattle.io/cn-myrke2-128-pool1-ef6ffd22-7jtc8:myrke2-128-pool1-ef6ffd22-7jtc8 listener.cattle.io/fingerprint:SHA1=83CDD81E2BC9612F9C7E7E06A70E367C36573FC7]"
    May 24 10:22:04 myrke2-128-pool1-ef6ffd22-7jtc8 rke2[2322]: time="2024-05-24T10:22:04Z" level=info msg="Active TLS secret kube-system/rke2-serving (ver=4210) (count 11): map[field.cattle.io/projectId:c-m-c79fjcww:p-gz6xh listener.cattle.io/cn-10.43.0.1:10.43.0.1 listener.cattle.io/cn-127.0.0.1:127.0.0.1 listener.cattle.io/cn-192.168.0.52:192.168.0.52 listener.cattle.io/cn-__1-f16284:::1 listener.cattle.io/cn-kubernetes:kubernetes listener.cattle.io/cn-kubernetes.default:kubernetes.default listener.cattle.io/cn-kubernetes.default.svc:kubernetes.default.svc listener.cattle.io/cn-kubernetes.default.svc.cluster.local:kubernetes.default.svc.cluster.local listener.cattle.io/cn-localhost:localhost listener.cattle.io/cn-myrke2-128-pool1-ef6ffd22-7jtc8:myrke2-128-pool1-ef6ffd22-7jtc8 listener.cattle.io/fingerprint:SHA1=83CDD81E2BC9612F9C7E7E06A70E367C36573FC7]"
    ```

3. `rancher-system-agent.service `
    ```
    # journalctl -u rancher-system-agent.service --follow
    May 24 10:34:11 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:34:11Z" level=info msg="[Applyinator] No image provided, creating empty working directory /var/lib/rancher/agent/work/20240524-103411/1a601d63558d61145f67d1569292e3450c7597bc81499ae31bebbec678d6a964_0"
    May 24 10:34:11 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:34:11Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
    May 24 10:34:12 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:34:12Z" level=info msg="[1a601d63558d61145f67d1569292e3450c7597bc81499ae31bebbec678d6a964_0:stdout]: Name Location Size Created"
    May 24 10:34:12 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:34:12Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 0"
    May 24 10:34:12 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:34:12Z" level=info msg="[K8s] updated plan secret fleet-default/myrke2-128-bootstrap-template-45xrz-machine-plan with feedback"
    May 24 10:44:13 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:44:13Z" level=info msg="[Applyinator] No image provided, creating empty working directory /var/lib/rancher/agent/work/20240524-104413/1a601d63558d61145f67d1569292e3450c7597bc81499ae31bebbec678d6a964_0"
    May 24 10:44:13 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:44:13Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
    May 24 10:44:13 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:44:13Z" level=info msg="[1a601d63558d61145f67d1569292e3450c7597bc81499ae31bebbec678d6a964_0:stdout]: Name Location Size Created"
    May 24 10:44:13 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:44:13Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 0"
    May 24 10:44:14 myrke2-128-pool1-ef6ffd22-7jtc8 rancher-system-agent[16919]: time="2024-05-24T10:44:14Z" level=info msg="[K8s] updated plan secret fleet-default/myrke2-128-bootstrap-template-45xrz-machine-plan with feedback"
    ```

**Comment by albinsun** [comment]
Same test with `rancher-v2.8.3` looks fine.
`0.2.300` > `103.0.1+up0.2.3` (latest)
![image](https://github.com/harvester/harvester/assets/2773781/501b4b50-3a8b-4d07-98f7-b657af20659a)

**Comment by albinsun** [comment]
One more trial.
* 2 options, we pick the latest (`103.0.2+up0.2.4`), so `0.2.300` > `103.0.2+up0.2.4`
  ![image](https://github.com/harvester/harvester/assets/2773781/dca0aa24-b195-4504-b05b-e90eb1263592)
* Fail, note that the pending failed pod uses `harvester-cloud-provider:v0.2.1` which is not the same as we specified (`v0.2.4`)
* Similiar to harvester/harvester/issues/5382
 ![image](https://github.com/harvester/harvester/assets/2773781/13f7bc32-a002-4d93-a306-83a952714780)

**Comment by bk201** [comment]
This should be duplicated with https://github.com/harvester/harvester/issues/5382, @starbops can you help check?

**Comment by khushboo-rancher** [comment]
Just leaving a workaround here - If we delete the old pod, the new pod with spin up successfully.

**Comment by starbops** [comment]
The root cause is the same as #5382's. The image repo/tag of `harvester-cloud-provider` was changed, which triggered the deployment rollout. Since it is a single-node guest cluster, the situation described in #5348 happened.

**Comment by albinsun** [comment]
Test PASS with Rancher `v2.8.5` and `harvester-cloud-provider-v0.2.5+`.

> [!NOTE]
> Note that this issue fixed in `harvester-cloud-provider:v0.2.5` which included in `rancher-v2.8+` but not `v2.7.x`. [^1]


## Environment
* `harvester-v1.4.0-rc3`
  * Profile: Single node AMD64 QEMU/KVM
  * ui-source: `Auto`
* `rancher-v2.8.5`
  * Profile: Single node AMD64 QEMU/KVM
  * K3s Channel: `v1.28`

## Prerequisite
1. Create image _ubuntu-22.04-jammy-server-cloudimg-amd64.img_
1. Create VM network _mgmt-vlan1_
1. <details>
   <summary> Import Harvester to Rancher
   </summary>

   ![image](https://github.com/user-attachments/assets/75fd1873-8931-4c21-9798-bd2b49213717)
   </details>


## Test Case
### CASE
1. <details>
   <summary> :heavy_check_mark: Create RKE2 cluster (<code>v1.28.13+rke2r1</code>)
   </summary>
   
   ![image](https://github.com/user-attachments/assets/20805641-9a59-4d7f-a65a-d4c6a47bf554)
   </details>

1. <details>
   <summary> :heavy_check_mark: Check <i>harvester-cloud-provider</i> chart version should less than <code>v0.2.5</code>
   </summary>
   
   ![image](https://github.com/user-attachments/assets/583f489a-9186-4d8d-890b-3b51787d4a51)
   </details>

1. <details>
   <summary> :heavy_check_mark: Deploy <i>mynginx1</i>
   </summary>
   
   ![image](https://github.com/user-attachments/assets/33f0ec75-a664-4e56-8acf-41d671001486)
   </details>

1. <details>
   <summary> :heavy_check_mark: Upgrade <i>harvester-cloud-provider</i> chart to <code>v0.2.5+</code>
   </summary>
   
   * Marketplace
     ![image](https://github.com/user-attachments/assets/b9ac3060-d850-4499-b6c8-3f09a95b5dce)
   * Upgrade
     ![image](https://github.com/user-attachments/assets/95723125-2550-4452-9f11-8a1fe1fedd53)
   </details>

1. <details>
   <summary> :heavy_check_mark: <i>mynginx1</i> still works,deploy <i>mynginx2</i>
   </summary>
   
   * mynginx1
     ![image](https://github.com/user-attachments/assets/7fe3cf57-eea0-4c69-930e-aed7bab2c8ef)
   * mynginx2
     ![image](https://github.com/user-attachments/assets/62b5153d-cbf4-4967-afcb-eaa5abf2528a)
   </details>


[^1]: https://github.com/rancher/charts/tree/release-v2.8/charts/harvester-cloud-provider
